rust/worker/src/execution/operators/merge_metadata_results.rs

```diff
@@ -107,289 +113,128 @@ impl Operator<MergeMetadataResultsOperatorInput, MergeMetadataResultsOperatorOut
-                            ),
-                        );
-                    }
-                };
+                error!("Error creating record segment reader {}", e);
+                return Err(MergeMetadataResultsOperatorError::RecordSegmentCreationError(*e));
             }
         };
 
-        // Step 1: Materialize the logs.
-        let materializer =
-            LogMaterializer::new(record_segment_reader, input.filtered_log.clone(), None);
-        let mat_records = match materializer
+        // Materialize the logs.
+        let materializer = LogMaterializer::new(
+            record_segment_reader.clone(),
+            input.filtered_log.clone(),
+            None,
+        );
+        let mat_records = materializer
             .materialize()
             .instrument(tracing::trace_span!(parent: Span::current(), "Materialize logs"))
             .await
-        {
-            Ok(records) => records,
-            Err(e) => {
-                return Err(MergeMetadataResultsOperatorError::LogMaterializationError(
-                    e,
+            .map_err(|e| {
+                error!("Error materializing log: {}", e);
+                MergeMetadataResultsOperatorError::LogMaterializationError(e)
+            })?;
+
+        // Merge the offset ids, assuming the user_offset_ids and filtered_offset_ids are ordered.
+        let mut merged_offset_ids = input
+            .user_offset_ids
+            .iter()
+            .chain(&input.filtered_offset_ids)
+            .map(Vec::clone)
+            .reduce(|user_ids, filter_ids| merge_sorted_vecs_conjunction(&user_ids, &filter_ids));
+
+        // If the merged offset ids is None, it suggests user did not specify any filter. We fetch all offset ids using the record segment reader
+        if merged_offset_ids.is_none() {
+            let mut log_offset_ids = mat_records
+                .iter()
+                .map(|(log, _)| log.offset_id)
+                .collect::<Vec<_>>();
+            log_offset_ids.sort();
+            if let Some(reader) = &record_segment_reader {
+                let compact_offset_ids = reader.get_all_offset_ids().await.map_err(|e| {
+                    error!("Error reading record segment: {}", e);
+                    MergeMetadataResultsOperatorError::RecordSegmentReadError
+                })?;
+                merged_offset_ids = Some(merge_sorted_vecs_disjunction(
+                    &log_offset_ids,
+                    &compact_offset_ids,
                 ));
             }
-        };
+        }
+
+        // Truncate the offset ids using offset and limit
+        merged_offset_ids = merged_offset_ids.map(|offset_ids| {
+            if input.offset.is_some() || input.limit.is_some() {
+                let skip_count = input.offset.map(|o| o as usize).unwrap_or(0);
+                let take_count = input.limit.map(|l| l as usize).unwrap_or(offset_ids.len());
+                offset_ids
```

Comment:
i think you can use Vec slicing instead of skip/take/collect. Its easier to read, and easier to know its performant?

Reply:
Refactored